crates/simulate/src/agent/agent_channel.rs

```diff
@@ -1,53 +1,138 @@
+            world_update_sender,
+            world_result_sender,
         }
     }
 
-    /// Sends an update to the stepper. If the tag of the AgentChannel is not none,
-    /// the update will be defaultly tagged with the AgentChannel tag.
-    pub fn send(&self, update: SimUpdate<U>) {
-        self.sender
-            .send(AgentSimUpdate {
-                update,
-                address: self.address,
-                tag: self.tag.clone().or(None),
-            })
-            .expect("Agent channel disconnected.");
+    pub fn execute_evm_tx_raw(&self, tx: TxEnv) {
+        self.evm_update_sender
+            .send(EvmStateUpdate::ExecuteRaw(
+                self.address,
+                tx,
+                self.evm_result_sender.clone(),
+            ))
+            .expect("Channel send error.");
     }
 
-    pub fn send_with_tag(&self, update: SimUpdate<U>, tag: Cow<'static, str>) {
-        self.sender
-            .send(AgentSimUpdate {
+    pub fn execute_evm_tx<D: Detokenize>(
+        &self,
+        call: ethers::contract::ContractCall<StateMiddleware, D>,
+    ) {
+        self.evm_update_sender
+            .send(EvmStateUpdate::Execute(
+                self.address,
+                call.tx,
+                call.function,
+                self.evm_result_sender.clone(),
+            ))
+            .expect("Channel send error.");
+    }
+
+    pub fn execute_world_update(&self, update: U) {
+        self.world_update_sender
+            .send(WorldStateUpdate {
                 update,
-                address: self.address,
-                tag: Some(tag),
+                result_sender: self.world_result_sender.clone(),
             })
-            .expect("Agent channel disconnected.");
+            .expect("Channel send error.");
+    }
+}
+
+pub struct AgentChannelReceiver<U: Update> {
+    pub evm_result_receiver: Receiver<EvmStateUpdateOutput>,
+    pub world_result_receiver: Receiver<U>,
+}
+
+impl<U: Update> AgentChannelReceiver<U> {
+    pub fn new(
+        evm_result_receiver: Receiver<EvmStateUpdateOutput>,
+        world_result_receiver: Receiver<U>,
+    ) -> Self {
+        Self {
+            evm_result_receiver,
+            world_result_receiver,
+        }
+    }
+
+    pub fn receive_evm_tx_output<D: Detokenize>(&self) -> AgentChannelResult<Option<D>> {
+        match self.evm_result_receiver.try_recv() {
+            Ok(output) => match output {
+                EvmStateUpdateOutput::Execute(evm_tx_output, function) => match evm_tx_output {
+                    EvmTxOutput::Success(output_bytes) => {
+                        let decoded = decode_output(&function, output_bytes)
+                            .map_err(StateError::FunctionOutputDecodeError)?;
+                        Ok(Some(decoded))
+                    }
+                    _ => Ok(None),
+                },
+                EvmStateUpdateOutput::ExecuteRaw(_) => Err(AgentChannelError::WrongEvmTxOutputType),
+            },
```

Comment:
For my own understanding, can you explain the steps occurring here?

Reply:
1. The first match statement checks if there are even any evm results in the channel. If not, it errors. That means there is some error in the agent logic since they should always get an update result for executed evm txs.
2. The second match statement matches on the type of the evm tx. You can either send a revm `TxEnv` directly or an ethers `ContractCall`. We can only decode into type `D` if you send a `ContractCall`, else you should receive using `receive_evm_tx_output_raw`.
3. Final inner match statement checks if the transaction itself was a success or not. If it was a success, decodes it for the agent into type `D` using the ethers function. If it was a revert, it just sends a None.